cpp/src/io/text/bgzip_data_chunk_source.cu

```diff
@@ -0,0 +1,444 @@
+        CUDF_EXPECTS(_stream->seekg(remaining_size - 6, std::ios_base::cur), "seek failed");
+        auto block_size = read_int<uint16_t>(&buffer[0]);
+        return {block_size + 1, extra_length};
+      } else {
+        CUDF_EXPECTS(_stream->seekg(subfield_size, std::ios_base::cur), "seek failed");
+        extra_offset += subfield_size;
+      }
+    }
+    CUDF_EXPECTS(false, "missing BGZIP size extra subfield");
+  }
+
+  struct bgzip_footer {
+    uint32_t decompressed_size;
+  };
+
+  bgzip_footer read_footer()
+  {
+    std::array<char, 8> buffer{};
+    CUDF_EXPECTS(_stream->read(buffer.data(), sizeof(buffer)), "read failed");
+    return {read_int<uint32_t>(&buffer[4])};
+  }
+
+  /*
+
+
+    High-level control flow
+
+
+  */
+
+  template <typename T>
+  using pinned_host_vector =
+    thrust::host_vector<T, thrust::system::cuda::experimental::pinned_allocator<T>>;
+
+  template <typename T>
+  static void copy_to_device(const pinned_host_vector<T>& host,
+                             rmm::device_uvector<T>& device,
+                             rmm::cuda_stream_view stream)
+  {
+    device.resize(host.size(), stream);
+    CUDF_CUDA_TRY(cudaMemcpyAsync(
+      device.data(), host.data(), host.size() * sizeof(T), cudaMemcpyHostToDevice, stream.value()));
+  }
+
+  struct compressed_blocks {
+    static constexpr std::size_t default_buffer_alloc =
+      1 << 24;  // 16MB buffer allocation, resized on demand
+    static constexpr std::size_t default_offset_alloc =
+      1 << 16;  // 64k offset allocation, resized on demand
+
+    cudaEvent_t event;
+    pinned_host_vector<char> h_compressed_blocks;
+    pinned_host_vector<std::size_t> h_compressed_offsets;
+    pinned_host_vector<std::size_t> h_decompressed_offsets;
+    rmm::device_uvector<char> d_compressed_blocks;
+    rmm::device_uvector<char> d_decompressed_blocks;
+    rmm::device_uvector<std::size_t> d_compressed_offsets;
+    rmm::device_uvector<std::size_t> d_decompressed_offsets;
+    rmm::device_uvector<device_span<const uint8_t>> d_compressed_spans;
+    rmm::device_uvector<device_span<uint8_t>> d_decompressed_spans;
+    rmm::device_uvector<decompress_status> d_decompressed_status;
+    std::size_t compressed_size_with_headers{};
+    std::size_t max_decompressed_size{};
+    // this is usually equal to decompressed_size()
+    // unless we are in the last chunk, where it's limited by _local_end
+    std::size_t available_decompressed_size{};
+    std::size_t read_pos{};
+    bool decompressed{};
+
+    compressed_blocks()
+      : d_compressed_blocks(0, cudf::default_stream_value),
+        d_decompressed_blocks(0, cudf::default_stream_value),
+        d_compressed_offsets(0, cudf::default_stream_value),
+        d_decompressed_offsets(0, cudf::default_stream_value),
+        d_compressed_spans(0, cudf::default_stream_value),
+        d_decompressed_spans(0, cudf::default_stream_value),
+        d_decompressed_status(0, cudf::default_stream_value)
```

Comment:
I feel `compressed_blocks` ctor could take a CUDA stream parameter and so does `reset()`.

Reply:
`compressed_blocks` doesn't do any device operations except for the decompress function, where it may reallocate etc. Due to the multistream setup, I wanted to isolate the stream-ordered operations to a single function. The constructor relies on empty `device_uvectors` pointing to `nullptr`, and `reset` relies on the `resize` function never shrinking allocations. Is that a safe assumption?

Reply:
Here I was mainly concerned by the way that we initialize each attribute with an individual default stream value. Passing in a stream parameter will make future extensions like using a different stream to construct `compressed_blocks` less laborious:
```cpp
    compressed_blocks(rmm::cuda_stream_view stream)
      : d_compressed_blocks(0, stream),
        d_decompressed_blocks(0, stream),
        d_compressed_offsets(0, stream),
        d_decompressed_offsets(0, stream),
        d_compressed_spans(0, stream),
        d_decompressed_spans(0, stream),
        d_decompressed_status(0, stream)
```

Reply:
makes sense, I'll make the change and explain why we can use the default stream there